File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int swapIndex = -1,n = nums.size();

        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                swapIndex = i;
                break;
            }
        }
            if(swapIndex!=-1){
            for(int i=n-1;i>=swapIndex;i--){
            if (nums[i] > nums[swapIndex] ) {
              swap(nums[i],nums[swapIndex]);
                break;
            }
        }
            }
    
    if (swapIndex != -1) reverse(nums.begin() + swapIndex+1, nums.end());
    else reverse(nums.begin(), nums.end());
}
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    void swap(int a,int b,int[] nums){
        int temp=nums[a];
        nums[a]=nums[b];
        nums[b]=temp;

    }

    void reverseArray(int startIndex,int endIndex,int[] nums){
            while(startIndex<endIndex){
                swap(startIndex,endIndex,nums);
                startIndex++;
                endIndex--;
            }
    }

    public void nextPermutation(int[] nums) {
        int n=nums.length, reverseIndex=-1;

        for(int i=n-1;i>=1;i--){
            if(nums[i-1]<nums[i]){
                reverseIndex=i-1;
                break;
            }
        }

        int swapIndex=0;

        if (reverseIndex != -1) {
            for (int i = n - 1; i > reverseIndex; i--) {
                if (nums[i] > nums[reverseIndex]) {
                    swap(i, reverseIndex, nums);
                    break;
                }
            }
        }

        reverseArray(reverseIndex+1,n-1,nums);
    }
}